src/osd/OSD.h

```diff
@@ -517,13 +521,29 @@ class OSDService {
     set<pair<utime_t,spg_t> >::iterator it = last_scrub_pg.find(p);
     assert(it != last_scrub_pg.end());
     last_scrub_pg.erase(it);
+    pg_reg_stamp.erase(pgid);
+    reject_scrub_pg.second = false;
   }
   bool first_scrub_stamp(pair<utime_t, spg_t> *out) {
-    Mutex::Locker l(sched_scrub_lock);
-    if (last_scrub_pg.empty())
+    sched_scrub_lock.Lock();
+    if (last_scrub_pg.empty()) {
+      sched_scrub_lock.Unlock();
       return false;
+    }
+    if (reject_scrub_pg.second) {
+      spg_t pgid = reject_scrub_pg.first;
+      reject_scrub_pg.second = false;
+      utime_t t = pg_reg_stamp[pgid];
+      pair<utime_t, spg_t> pos(t, pgid);
+      sched_scrub_lock.Unlock();
+      if (next_scrub_stamp(pos, out)) {
```

Comment:
Nope, if you want to do this, refactor next_scrub_stamp into

next_scrub_stamp(...) {
  Mutex::Locker l(sched_scrub_lock)
  return _next_scrub_stamp();
}

It's not worth complicating the locking here to avoid a helper method.